src/core/reference/include/openvino/reference/adaptive_rkv_diversity.hpp

```diff
@@ -0,0 +1,225 @@
+                              /* transpose_arg1 = */ true);
+        normalized_key_data_buf.reset();
+
+        Shape evictable_subset_shape = {key_shape[0], m_eviction_size, m_eviction_size};
+        auto evictable_subset_buf = allocate_buf(evictable_subset_shape);
+        // stops?
+        ov::reference::slice(reinterpret_cast<char*>(cos_similar_buf.get()),
+                             cos_similar_shape,
+                             reinterpret_cast<char*>(evictable_subset_buf.get()),
+                             evictable_subset_shape,
+                             sizeof(T),
+                             /* starts = */ {m_start_size, m_start_size},
+                             /* steps = */ {1, 1},
+                             /* axes = */ {1, 2});
+        cos_similar_buf.reset();
+
+        fill_diagonal_(evictable_subset_buf.get(), evictable_subset_shape, 0.0);
+
+        Shape means_shape = {key_shape[0], m_eviction_size};
+        auto means_buf = allocate_buf(means_shape);
+        ov::reference::reduce_mean(evictable_subset_buf.get(), means_buf.get(), evictable_subset_shape, {2});
+
+        fill_low_values_with_zeros_(evictable_subset_buf.get(), evictable_subset_shape, means_buf.get(), means_shape);
+        means_buf.reset();
+
+        Shape aggregated_token_similarities_shape = {m_eviction_size, m_eviction_size};
+        auto aggregated_token_similarities_buf = allocate_buf(aggregated_token_similarities_shape);
+        ov::reference::reduce_mean(evictable_subset_buf.get(),
+                                   aggregated_token_similarities_buf.get(),
+                                   evictable_subset_shape,
+                                   {0});
+        evictable_subset_buf.reset();
+
+        Shape block_diversity_shape = {m_eviction_size / m_block_size, m_eviction_size};
+        auto block_diversity_buf = allocate_buf(block_diversity_shape);
+        block_sum_diversity_values(aggregated_token_similarities_buf.get(),
+                                   aggregated_token_similarities_shape,
+                                   block_diversity_buf.get(),
+                                   block_diversity_shape);
+        std::vector<std::vector<T>> retval(block_diversity_shape[0], std::vector<T>(block_diversity_shape[1]));
+        for (size_t block_idx = 0; block_idx < block_diversity_shape[0]; block_idx++) {
+            for (size_t token_idx = 0; token_idx < block_diversity_shape[1]; token_idx++) {
+                retval[block_idx][token_idx] = block_diversity_buf[block_idx * block_diversity_shape[1] + token_idx];
+            }
+        }
+
+        return retval;
+    }
+
+    /**
+     * @param shape Shape of a tensor
+     * @return A shared_ptr owning a buffer that can be used to store tensor data for the given shape.
+     * */
+    std::shared_ptr<T[]> allocate_buf(const Shape& shape) {
+        return std::shared_ptr<T[]>(new T[ov::shape_size(shape)]);
+    }
```

Comment:
Can `std::shared_ptr<T[]>` be replaced with` std::vector<T>` to simplify the code?